lib/handlers/authenticate-handler.ts

```diff
@@ -0,0 +1,273 @@
+import {
+  InsufficientScopeError,
+  InvalidArgumentError,
+  InvalidRequestError,
+  InvalidTokenError,
+  OAuthError,
+  ServerError,
+  UnauthorizedRequestError,
+} from '../errors';
+import { Model, Token } from '../interfaces';
+import { Request } from '../request';
+import { Response } from '../response';
+
+export class AuthenticateHandler {
+  addAcceptedScopesHeader: any;
+  addAuthorizedScopesHeader: any;
+  allowBearerTokensInQueryString: any;
+  model: Model;
+  scope: any;
+  constructor(options: any = {}) {
+    if (!options.model) {
+      throw new InvalidArgumentError('Missing parameter: `model`');
+    }
+
+    if (!options.model.getAccessToken) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `getAccessToken()`',
+      );
+    }
+
+    if (options.scope && options.addAcceptedScopesHeader === undefined) {
+      throw new InvalidArgumentError(
+        'Missing parameter: `addAcceptedScopesHeader`',
+      );
+    }
+
+    if (options.scope && options.addAuthorizedScopesHeader === undefined) {
+      throw new InvalidArgumentError(
+        'Missing parameter: `addAuthorizedScopesHeader`',
+      );
+    }
+
+    if (options.scope && !options.model.verifyScope) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `verifyScope()`',
+      );
+    }
+
+    this.addAcceptedScopesHeader = options.addAcceptedScopesHeader;
+    this.addAuthorizedScopesHeader = options.addAuthorizedScopesHeader;
+    this.allowBearerTokensInQueryString =
+      options.allowBearerTokensInQueryString;
+    this.model = options.model;
+    this.scope = options.scope;
+  }
+
+  /**
+   * Authenticate Handler.
+   */
+
+  async handle(request: Request, response: Response) {
+    if (!(request instanceof Request)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `request` must be an instance of Request',
+      );
+    }
+
+    if (!(response instanceof Response)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `response` must be an instance of Response',
+      );
+    }
+
+    // Extend model object with request
+    this.model.request = request;
```

Comment:
This is scary: Two requests in parallel could produce quite some confusion here it seems. The model is shared between all handlers (and whatever else gets it through the options), and only passed by reference. 

(This is a problem also in the original code)

Reply:
We should remove the request from the model and see if something breaks. I think it is only bad if you really access it.